google-services-plugin/src/main/kotlin/com/google/gms/googleservices/GoogleServicesPlugin.kt

```diff
@@ -0,0 +1,153 @@
+            )
+            project.configurations.all { projectConfig ->
+                if (projectConfig.name.contains("ompile")) {
+                    projectConfig.incoming.afterResolve(strictVersionDepInspector::afterResolve)
+                }
+            }
+        }
+
+        var pluginApplied = false
+
+        project.pluginManager.withPlugin("com.android.application") {
+            pluginApplied = true
+            project.extensions.configure(ApplicationAndroidComponentsExtension::class.java) {
+                it.registerSourceType(SOURCE_TYPE)
+                it.onVariants { variant ->
+                    handleVariant(variant, project)
+                }
+            }
+        }
+        project.pluginManager.withPlugin("com.android.dynamic-feature") {
+            pluginApplied = true
+            project.extensions.configure(DynamicFeatureAndroidComponentsExtension::class.java) {
+                it.registerSourceType(SOURCE_TYPE)
+                it.onVariants { variant ->
+                    handleVariant(variant, project)
+                }
+            }
+        }
+
+        project.afterEvaluate {
+            if (pluginApplied) {
+                return@afterEvaluate
+            }
+            project.logger.error(
+                "The google-services Gradle plugin needs to be applied on a project with " +
+                        "com.android.application or com.android.dynamic-feature."
+            )
+        }
+    }
+
+    private fun <T> handleVariant(
+        variant: T,
+        project: Project
+    ) where T : Variant, T : GeneratesApk {
+        val jsonToXmlTask = project.tasks.register(
+            "process${variant.name.capitalize()}GoogleServices",
+            GoogleServicesTask::class.java
+        ) {
+            it.missingGoogleServicesStrategy.set(
+                project.extensions.getByType(GoogleServicesPluginConfig::class.java)
+                    .missingGoogleServicesStrategy
+            )
+            it.googleServicesJsonFiles.set(variant.sources.getByName(SOURCE_TYPE).all)
+            it.applicationId.set(variant.applicationId)
+        }
+
+        // TODO: add an AGP version check to this block
+        //  when https://issuetracker.google.com/issues/268192807 is fixed
+        try {
+            variant.sources.getByName(SOURCE_TYPE)
+                .addStaticSourceDirectory(
+                    project
+                        .layout
+                        .projectDirectory
+                        .dir("src/${variant.name}/google-services").toString()
+                )
+        } catch (e: IllegalArgumentException) {
+            // directory doesn't exist, ignore
+        }
+        variant.sources.res?.addGeneratedSourceDirectory(
+            jsonToXmlTask,
+            GoogleServicesTask::outputDirectory
+        )
+    }
+
+    /* Recommended replacement for Kotlin's deprecated capitalize function */
+    private fun String.capitalize(): String =
```

Comment:
Gradle actually has an extension method for this: 
```kotlin
import org.gradle.configurationcache.extensions.capitalized
```

Reply:
that looks like it is public by accident and is not Gradle's public API (https://docs.gradle.org/current/javadoc/index.html)

I'd be afraid to use it in case it is changed/deleted across Gradle versions, but if you still prefer it I'll make the change